File: advent_of_code/2024/day1/zig/src/main.zig
const std = @import("std");

pub fn part1(input: []u8) !i32 {
    var dsum: i32 = 0;
    const allocator = std.heap.page_allocator;
    var left = std.ArrayList(i32).init(allocator);
    defer left.deinit();

    var right = std.ArrayList(i32).init(allocator);
    defer right.deinit();

    var lines = std.mem.splitSequence(u8, input, "\n");

    while (lines.next()) |line| {
        var tokens = std.mem.splitSequence(u8, line, "   ");
        var token = tokens.next().?;
        if (token.len == 0) {
            break;
        }
        std.debug.print("Token: {s}\n", .{token});
        try left.append(try std.fmt.parseInt(i32, token, 10));
        token = tokens.next().?;
        std.debug.print("Token: {s}\n", .{token});
        try right.append(try std.fmt.parseInt(i32, token, 10));
    }

    std.mem.sort(i32, left.items[0..], {}, comptime std.sort.asc(i32));
    std.mem.sort(i32, right.items[0..], {}, comptime std.sort.asc(i32));

    for (left.items, 0..) |_, i| {
        const d: u32 = @abs(left.items[i] - right.items[i]);
        dsum += @intCast(d);
    }

    return dsum;
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const input = try std.fs.cwd().openFile("sample.txt", .{});
    defer input.close();

    const input_size = try input.getEndPos();
    const input_buffer = try allocator.alloc(u8, input_size);
    defer allocator.free(input_buffer);

    _ = try input.readAll(input_buffer);

    const result = try part1(input_buffer);
    std.debug.print("Result: {}\n", .{result});
}

File: advent_of_code/2024/day1/go/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"sort"
	"strconv"
	"strings"
)

func part1(input string) int {
	dsum := 0
	lines := strings.Split(input, "\n")

	left := []int{}
	right := []int{}

	for i, line := range lines {
		tokens := strings.SplitN(line, "   ", 2)
		if len(tokens) != 2 {
			continue
		}

		num, err := strconv.Atoi(tokens[0])
		if err != nil {
			log.Fatalf("Invalid number at line %d: %s", i+1, tokens[0])
			continue
		}
		left = append(left, num)

		num, err = strconv.Atoi(tokens[1])
		if err != nil {
			log.Fatalf("Invalid number at line %d: %s", i+1, tokens[1])
			continue
		}
		right = append(right, num)
	}

	sort.Ints(left)
	sort.Ints(right)

	for i := range left {
		diff := math.Abs(float64(left[i] - right[i]))
		dsum += int(diff)
	}

	return dsum
}

func part2(input string) int {
	lines := strings.Split(input, "\n")
	left := []int{}
	right := make(map[int]int)
	score := 0

	for i, line := range lines {
		tokens := strings.SplitN(line, "   ", 2)
		if len(tokens) != 2 {
			continue
		}

		num, err := strconv.Atoi(tokens[0])
		if err != nil {
			log.Fatalf("Invalid number at line %d: %s", i+1, tokens[0])
		}
		left = append(left, num)

		num, err = strconv.Atoi(tokens[1])
		if err != nil {
			log.Fatalf("Invalid number at line %d: %s", i+1, tokens[1])
		}
		right[num]++
	}

	for _, num := range left {
		score += num * right[num]
	}

	return score
}

func main() {
	filePath := "sample.txt"

	data, err := ioutil.ReadFile(filePath)
	if err != nil {
		log.Fatalf("Error reading file: %v", err)
	}

	part1 := part1(string(data))
	fmt.Printf("Part 1: %d\n", part1)

	part2 := part2(string(data))
	fmt.Printf("Part 2: %d\n", part2)
}
